DPGAnalysis/Skims/src/ZElectronsSelector.cc

```diff
@@ -0,0 +1,197 @@
+  bool operator()(const reco::GsfElectron & ) const;
+  void newEvent (const edm::Event&, const edm::EventSetup&);
+  const float getEffectiveArea(float eta) const;
+  void printEffectiveAreas() const;
+
+  edm::EDGetTokenT<double> theRhoToken;
+  edm::EDGetTokenT<reco::GsfElectronCollection> theGsfEToken;
+  edm::Handle< double > _rhoHandle;
+
+  std::vector<double> absEtaMin_; // low limit of the eta range
+  std::vector<double> absEtaMax_; // upper limit of the eta range
+  std::vector<double> effectiveAreaValues_; // effective area for this eta range
+
+  edm::ParameterSet eleIDWP;
+
+  vector<int> missHits ;
+  vector<double> sigmaIEtaIEtaCut;
+  vector<double> dEtaInSeedCut   ;
+  vector<double> dPhiInCut       ;
+  vector<double> hOverECut       ;
+  vector<double> relCombIso      ;
+  vector<double> EInvMinusPInv   ;
+
+
+};
+
+
+void ZElectronsSelector::printEffectiveAreas() const {
+ 
+ 
+   printf("  eta_min   eta_max    effective area\n");
+   uint nEtaBins = absEtaMin_.size();
+   for(uint iEta = 0; iEta<nEtaBins; iEta++){
+     printf("  %8.4f    %8.4f   %8.5f\n",
+        absEtaMin_[iEta], absEtaMax_[iEta],
+        effectiveAreaValues_[iEta]);
+   }
+ 
+}
+const float ZElectronsSelector::getEffectiveArea(float eta) const{
+
+  float effArea = 0;
+  uint nEtaBins = absEtaMin_.size();
+  for(uint iEta = 0; iEta<nEtaBins; iEta++){
+    if( std::abs(eta) >= absEtaMin_[iEta]
+    && std::abs(eta) < absEtaMax_[iEta] ){
+      effArea = effectiveAreaValues_[iEta];
+      break;
+    }
+  }
+
+  return effArea;
+}
+
+
+ZElectronsSelector::ZElectronsSelector(const edm::ParameterSet& cfg, edm::ConsumesCollector & iC ) :
+  theRhoToken(iC.consumes <double> (cfg.getParameter<edm::InputTag>("rho")))
+{
+  absEtaMin_ = cfg.getParameter<std::vector<double> >("absEtaMin");
+  absEtaMax_ = cfg.getParameter<std::vector<double> >("absEtaMax");
+  effectiveAreaValues_= cfg.getParameter<std::vector<double> >("effectiveAreaValues"); 
+  //printEffectiveAreas();
+  eleIDWP = cfg.getParameter<edm::ParameterSet>("eleID");
+
+  missHits = eleIDWP.getParameter<std::vector<int> >("missingHitsCut");
+  sigmaIEtaIEtaCut= eleIDWP.getParameter<std::vector<double> >("full5x5_sigmaIEtaIEtaCut");
+  dEtaInSeedCut   = eleIDWP.getParameter<std::vector<double> >("dEtaInSeedCut");
+  dPhiInCut       = eleIDWP.getParameter<std::vector<double> >("dPhiInCut");
+  hOverECut       = eleIDWP.getParameter<std::vector<double> >("hOverECut");
+  relCombIso      = eleIDWP.getParameter<std::vector<double> >("relCombIsolationWithEACut");
+  EInvMinusPInv   = eleIDWP.getParameter<std::vector<double> >("EInverseMinusPInverseCut");
+}
+
+
+
+void  ZElectronsSelector::newEvent(const edm::Event& ev, const edm::EventSetup& ){
+
+  ev.getByToken(theRhoToken,_rhoHandle);
+  
+}
+
+bool ZElectronsSelector::operator()(const reco::GsfElectron& el) const{
+
+  bool keepEle = false; 
+  float pt_e = el.pt();
+
+  auto etrack  = el.gsfTrack().get(); 
+
+  if (el.isEB()){
```

Comment:
hi @taroni - please simplify this if statement by first setting the index value - eg
unsisnged int ind=0;
if ( ! el.isEB()) ind=1; //and add the eta requirements here  
then use ind in each of the checks instead of [0] or [1]

Reply:
Changed according to your suggestion